units/en/unit5/2_first-agent-interface.mdx

```diff
@@ -0,0 +1,228 @@
+                            }
+                    )
+                    messages_to_display.append(tool_message)
+                    yield messages_to_display
+                    
+        # Handle tool responses
+        if chunk.get("tools"):
+            for tool_response in chunk["tools"]["messages"]:
+                # Find the corresponding tool message
+                for msg in messages_to_display:
+                    if msg.metadata.get("id") == tool_response.tool_call_id:
+                        msg.content += f"\nResult: {tool_response.content}"
+                        yield messages_to_display
+    
+    # Add the final response as a regular message
+    if final_response:
+        messages_to_display.append(ChatMessage(content=final_response))
+        yield messages_to_display
+
+# Create the Gradio interface
+demo = gr.ChatInterface(
+    fn=stream_from_agent,
+    type="messages",
+    title="🌤️ Weather Assistant",
+    description="Ask about the weather anywhere! Watch as I gather the information step by step.",
+    examples=[
+        "What's the weather like in Tokyo?",
+        "Is it sunny in Paris right now?",
+        "Should I bring an umbrella in New York today?"
+    ],
+
+)
+
+demo.launch()
+```
+
+This creates a complete chat interface:
+<add an image or a gif of the chatinterface. It is available at the location: https://huggingface.co/spaces/ysharma/Your_First_Agent_UI. Note that we first need to set OpenAI API key as an env variable oin the Space>
+
+## Understanding the key elements
+
+Let's break down what's happening:
+
+1. **The Agent Function**:
+   ```python
+   def stream_from_agent(message: str, history: List[Dict[str, str]]) -> gr.ChatMessage:
+   ```
+   - Processes messages through the LangChain agent
+   - Takes the current message and chat history
+   - Uses `yield` to show intermediate tool-usage and thoughts
+   - Returns the final response
+
+2. **Tools available**:
+   ```python
+   def get_lat_lng(location_description: str) -> dict[str, float]:
+   def get_weather(lat: float, lng: float) -> dict[str, str]:
+   ```
+   - get_lat_lng : Get lattitude and longitude of a place from the place description
+   - get_weather : Get weather details for the given latitude and logitude 
+
+3. **ChatInterface Configuration**:
+   ```python
+    demo = gr.ChatInterface(
+        fn=stream_from_agent,
+        title="🌤️ Weather Assistant",
+       ...
+   )
+   ```
+   - Connects the Agent function
+   - Sets up the UI appearance
+   - Configures interactive features
+   <I can add more latest features to the chatbot (like history, sidebar etc.)>
+
+3. **Streaming Responses**:
+   - Using `yield` enables real-time updates
+   - Shows the Agent's thought process and tool-usage in real time
+   - Keeps users engaged while waiting
+
+We are using `type="messages"` in gr.ChatInterface. This passes the history as a list of dictionaries with openai-style "role" and "content" keys.
```

Comment:
We should talk about the `gr.ChatMessage` class briefly here and explain that we'll dive more deeply in the next section